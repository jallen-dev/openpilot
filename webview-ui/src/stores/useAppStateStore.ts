import { create } from "zustand"

interface AppState {
  workspaceIndexed: boolean
  setWorkspaceIndexed: (workspaceIndexed: boolean) => void
}

export const useAppStateStore = create<AppState>()((set, get) => ({
  workspaceIndexed: false,
  setWorkspaceIndexed: (workspaceIndexed: boolean) => {
    set({ workspaceIndexed })
  }
}))
