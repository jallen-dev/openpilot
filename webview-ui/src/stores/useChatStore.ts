import { create } from "zustand"
import { immer } from "zustand/middleware/immer"

import { ChatMessage } from "../../../src/shared/ChatMessage"
import { vscode } from "../utils/vscode"

interface ChatState {
  userPrompt: string
  loading: boolean // used to show loading indicator for non-streaming chats
  chatMessages: ChatMessage[]
  setUserPrompt: (prompt: string) => void
  setLoading: (loading: boolean) => void
  setChatMessages: (messages: ChatMessage[]) => void
}

export const useChatStore = create<ChatState>()(
  immer((set, get) => ({
    userPrompt: vscode.getState()?.userPrompt ?? "",
    chatMessages: vscode.getState()?.chatMessages ?? [],
    loading: false,
    setUserPrompt: (userPrompt: string) => {
      vscode.setState({ userPrompt })
      set({ userPrompt })
    },
    setLoading: (loading: boolean) => set({ loading }),
    setChatMessages: (chatMessages: ChatMessage[]) => {
      vscode.setState({ chatMessages })
      set({ chatMessages })
    }
  }))
)
