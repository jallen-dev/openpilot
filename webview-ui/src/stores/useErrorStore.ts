import { create } from "zustand"

import { vscode } from "../utils/vscode"

interface ErrorState {
  errorMessage: string
  setErrorMessage: (errorMessage: string) => void
}

export const useErrorStore = create<ErrorState>()((set, get) => ({
  errorMessage: vscode.getState()?.errorMessage ?? "",
  setErrorMessage: (errorMessage: string) => {
    vscode.setState({ errorMessage })
    set({ errorMessage })
  }
}))
