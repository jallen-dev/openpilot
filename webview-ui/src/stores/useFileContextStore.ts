import { create } from "zustand"

import { File, FileSearchResult } from "../../../src/shared/File"
import { vscode } from "../utils/vscode"

interface FileContextState {
  suggestedFiles: FileSearchResult[]
  rejectedFilePaths: string[]
  setSuggestedFiles: (files: FileSearchResult[]) => void
  rejectFile: (file: string) => void
  undoRejectFile: (file: string) => void
  clearRejectedFiles: () => void
  getFilesToInclude: () => File[]
}

export const useFileContextStore = create<FileContextState>()((set, get) => ({
  suggestedFiles: vscode.getState()?.suggestedFiles ?? [],
  rejectedFilePaths: vscode.getState()?.rejectedFilePaths ?? [],
  setSuggestedFiles: (suggestedFiles: FileSearchResult[]) => {
    vscode.setState({ suggestedFiles })
    set({ suggestedFiles })
  },
  rejectFile: (file: string) =>
    set((state) => {
      vscode.setState({ rejectedFilePaths: [...state.rejectedFilePaths, file] })
      return {
        rejectedFilePaths: [...state.rejectedFilePaths, file]
      }
    }),
  undoRejectFile: (file: string) =>
    set((state) => {
      vscode.setState({
        rejectedFilePaths: state.rejectedFilePaths.filter((f) => f !== file)
      })
      return {
        rejectedFilePaths: state.rejectedFilePaths.filter((f) => f !== file)
      }
    }),
  clearRejectedFiles: () => {
    vscode.setState({ rejectedFilePaths: [] })
    set({ rejectedFilePaths: [] })
  },
  getFilesToInclude: () => {
    const { suggestedFiles, rejectedFilePaths } = get()
    return suggestedFiles.filter((f) => !rejectedFilePaths.includes(f.path))
  }
}))
