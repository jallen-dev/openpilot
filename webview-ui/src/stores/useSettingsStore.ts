import { create } from "zustand"

import { ApiKeys } from "../../../src/shared/ApiKeys"
import { Configuration } from "../../../src/shared/Configuration"
import { Model } from "../../../src/shared/Model"

interface SettingsState {
  model: Model
  apiKeys: ApiKeys
  includeFiles: boolean
  produceDiffs: boolean
  setModel: (model: Model) => void
  setApiKeys: (apiKeys: ApiKeys) => void
  setIncludeFiles: (includeFiles: boolean) => void
  setProduceDiffs: (produceDiffs: boolean) => void
  setConfiguration: (configuration: Configuration) => void
}

export const useSettingsStore = create<SettingsState>()((set, get) => ({
  model: "gpt-3.5-turbo",
  apiKeys: {},
  includeFiles: false,
  produceDiffs: false,
  setModel: (newModel) => set({ model: newModel }),
  setApiKeys: (newKeys) => {
    const { apiKeys } = get()
    set({ apiKeys: { ...apiKeys, ...newKeys } })
  },
  setIncludeFiles: (includeFiles) => set({ includeFiles }),
  setProduceDiffs: (produceDiffs) => set({ produceDiffs }),
  setConfiguration: (configuration: Configuration) => {
    const { includeFiles, produceDiffs } = configuration
    set({ includeFiles, produceDiffs })
  }
}))
