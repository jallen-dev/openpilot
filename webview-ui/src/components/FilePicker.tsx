import { VSCodeCheckbox } from "@vscode/webview-ui-toolkit/react"

import { MatchingFilesMessage } from "../../../src/shared/OpenpilotMessage"
import useMessageListener from "../hooks/useMessageListener"
import { useFileContextStore } from "../stores/useFileContextStore"
import { useSettingsStore } from "../stores/useSettingsStore"
import { truncate } from "../utils/truncate"
import { vscode } from "../utils/vscode"

function FilePicker() {
  const suggestedFiles = useFileContextStore((state) => state.suggestedFiles)
  const rejectedFilePaths = useFileContextStore(
    (state) => state.rejectedFilePaths
  )

  const setSuggestedFiles = useFileContextStore(
    (state) => state.setSuggestedFiles
  )
  const rejectFile = useFileContextStore((state) => state.rejectFile)
  const undoRejectFile = useFileContextStore((state) => state.undoRejectFile)

  const includeFiles = useSettingsStore((state) => state.includeFiles)

  useMessageListener<MatchingFilesMessage>("matchingFiles", (message) => {
    const files = message.files
    if (files) {
      setSuggestedFiles(files)
    }
  })

  if (!includeFiles) {
    return null
  }

  if (suggestedFiles.length === 0) {
    return null
  }

  const showFile = (path: string) => {
    vscode.postMessage({
      type: "showFile",
      path
    })
  }

  return (
    <div className="flex flex-col gap-2 py-4">
      <span>Files to include:</span>
      {suggestedFiles
        .sort((a, b) => a.score - b.score)
        .map((file) => (
          <div key={file.path} className="flex flex-row gap-4">
            <VSCodeCheckbox
              checked={!rejectedFilePaths.includes(file.path)}
              onChange={() =>
                rejectedFilePaths.includes(file.path)
                  ? undoRejectFile(file.path)
                  : rejectFile(file.path)
              }
            />
            <div className="grow">
              <a href="#" onClick={() => showFile(file.path)} title={file.path}>
                {truncate(file.path, 40)}
              </a>
            </div>
          </div>
        ))}
    </div>
  )
}

export default FilePicker
