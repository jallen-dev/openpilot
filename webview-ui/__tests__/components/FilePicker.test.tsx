import { render, screen } from "@testing-library/react"

import FilePicker from "../../src/components/FilePicker"
import { useFileContextStore } from "../../src/stores/useFileContextStore"
import { useSettingsStore } from "../../src/stores/useSettingsStore"

const testFileSearchResults = [
  { path: "src/file1.ts", name: "file1.ts", content: "content1", score: 0.5 },
  { path: "src/file2.ts", name: "file1.ts", content: "content2", score: 0.5 }
]

test("renders nothing when includeFiles is false", () => {
  useSettingsStore.getState().setIncludeFiles(false)
  useFileContextStore.getState().setSuggestedFiles(testFileSearchResults)
  render(<FilePicker />)

  expect(screen.queryByText(/Files to include/i)).not.toBeInTheDocument()
})

test("renders nothing when there are no files", () => {
  useSettingsStore.getState().setIncludeFiles(false)
  render(<FilePicker />)

  expect(screen.queryByText(/Files to include/i)).not.toBeInTheDocument()
})

test("renders a list of files", () => {
  useSettingsStore.getState().setIncludeFiles(true)
  useFileContextStore.getState().setSuggestedFiles(testFileSearchResults)
  render(<FilePicker />)

  expect(screen.queryByText(/Files to include/i)).toBeInTheDocument()
})

test("matches snapshot", () => {
  useSettingsStore.getState().setIncludeFiles(true)
  useFileContextStore.getState().setSuggestedFiles(testFileSearchResults)
  const { asFragment } = render(<FilePicker />)

  expect(asFragment()).toMatchSnapshot()
})
