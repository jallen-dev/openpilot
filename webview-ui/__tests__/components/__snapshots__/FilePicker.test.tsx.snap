// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col gap-2 py-4"
  >
    <span>
      Files to include:
    </span>
    <div
      class="flex flex-row gap-4"
    >
      <vscode-checkbox
        aria-checked="false"
        aria-disabled="false"
        aria-label="Checkbox"
        aria-required="false"
        role="checkbox"
        tabindex="0"
      >
        <input
          slot="form-associated-proxy"
          style="display: none;"
          type="checkbox"
          value="on"
        />
      </vscode-checkbox>
      <div
        class="grow"
      >
        <a
          href="#"
          title="src/file1.ts"
        >
          src/file1.ts
        </a>
      </div>
    </div>
    <div
      class="flex flex-row gap-4"
    >
      <vscode-checkbox
        aria-checked="false"
        aria-disabled="false"
        aria-label="Checkbox"
        aria-required="false"
        role="checkbox"
        tabindex="0"
      >
        <input
          slot="form-associated-proxy"
          style="display: none;"
          type="checkbox"
          value="on"
        />
      </vscode-checkbox>
      <div
        class="grow"
      >
        <a
          href="#"
          title="src/file2.ts"
        >
          src/file2.ts
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
